src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java

```diff
@@ -52,4 +93,278 @@ private long countGermplasm(UUID programId, String trialDbId) throws ApiExceptio
+        if ((StringUtils.isBlank(params.getDataset()) || "observations".equalsIgnoreCase(params.getDataset())) &&
+                experiment.getAdditionalInfo().getAsJsonObject().get(BrAPIAdditionalInfoFields.OBSERVATION_DATASET_ID) != null) {
+            String obsDatasetId = experiment
+                    .getAdditionalInfo().getAsJsonObject()
+                    .get(BrAPIAdditionalInfoFields.OBSERVATION_DATASET_ID).getAsString();
+            isDataset = true;
+            obsVars = getDatasetObsVars(obsDatasetId, program);
+            addObsVarColumns(columns, obsVars, params.isIncludeTimestamps(), program);
+
+        }
+
+        // make export rows from any observations
+        if (isDataset) {
+            dataset = observationDAO.getObservationsByTrialDbId(List.of(experiment.getTrialDbId()), program);
+        }
+        if (!requestedEnvNames.isEmpty()) {
+            dataset = filterDatasetByEnvironment(dataset, requestedEnvNames);
+        }
+        addBrAPIObsToRecords(dataset, experiment, program, ous, studyByDbId, rowByOUId, params.isIncludeTimestamps(), obsVars);
+
+        // make export rows for OUs without observations
+        if (rowByOUId.size() < ous.size()) {
+            for (BrAPIObservationUnit ou: ous) {
+                String ouId = getOUId(ou);
+                if (!rowByOUId.containsKey(ouId)) {
+                    rowByOUId.put(ouId, createExportRow(experiment, program, ou, studyByDbId));
+                }
+            }
+        }
+
+        // write export data to requested file format
+        List<Map<String, Object>> exportRows = new ArrayList<>(rowByOUId.values());
+        if (fileType.equals(FileType.CSV)){
+            downloadFile = CSVWriter.writeToDownload(columns, exportRows, fileType);
+        } else {
+            downloadFile = ExcelWriter.writeToDownload("Dataset Export", columns, exportRows, fileType);
+        }
+
+        String envFilenameFragment = params.getEnvironments() == null ? "All Environments" : params.getEnvironments();
+        String fileName = makeFileName(experiment, program, envFilenameFragment) + fileType.getExtension();
+        return new DownloadFile(fileName, downloadFile);
+    }
+
+    private void addBrAPIObsToRecords(
+            List<BrAPIObservation> dataset,
+            BrAPITrial experiment,
+            Program program,
+            List<BrAPIObservationUnit> ous,
+            Map<String, BrAPIStudy> studyByDbId,
+            Map<String, Map<String, Object>> rowByOUId,
+            boolean includeTimestamp,
+            List<BrAPIObservationVariable> obsVars) throws ApiException, DoesNotExistException {
+        for (BrAPIObservation obs: dataset) {
+            BrAPIObservationUnit ou = ous.stream()
+                    .filter(unit -> obs.getObservationUnitDbId().equals(unit.getObservationUnitDbId()))
+                    .findAny()
+                    .orElseThrow(RuntimeException::new);
```

Comment:
I think it would be worth creating a map outside of the `for` loop of `ouByOUId` to reduce the runtime complexity

Reply:
done

Reply:
done